onnxruntime/core/framework/stream_execution_context.cc

```diff
@@ -0,0 +1,289 @@
+    release_plan_[i] = static_cast<int>(release_actions[i].ref_count);
+  }
+}
+
+synchronize::Notification* StreamExecutionContext ::GetNotification(size_t /*idx*/) { 
+    ORT_THROW("Try to get notification in a build which doesn't enable Stream!"); 
+}
+
+bool StreamExecutionContext ::DecCountDownBarrier(size_t /*barrier_id*/) {
+    ORT_THROW("Try to decrease barrier in a build which doesn't enable Stream!");
+}
+
+Stream* StreamExecutionContext ::GetDeviceStream(size_t /*idx*/) {
+    return nullptr;
+}
+#endif
+
+const SessionState& StreamExecutionContext ::GetSessionState() const { return *session_state_; }
+
+const logging::Logger& StreamExecutionContext ::GetLogger() const { return *logger_; }
+
+ExecutionFrame& StreamExecutionContext ::GetExecutionFrame() { return frame_; }
+
+const Status& StreamExecutionContext ::TaskStatus() const {
+  return task_status_;
+}
+
+void StreamExecutionContext ::CompleteTask() {
+  remain_tasks_.Dec();
+}
+
+void StreamExecutionContext ::AddTask() {
+  remain_tasks_.Inc();
+}
+
+void StreamExecutionContext ::WaitAll() {
+  while (remain_tasks_.Get())
+    onnxruntime::concurrency::SpinPause();
+}
+
+void StreamExecutionContext ::SetStatus(Status& status) {
+  // TODO: if multiple worker report non-ok status,
+  // what is our strategy? currently we just keep
+  // a random one. as long as it is not OK, the
+  // execution will fail.
+  if (task_status_.IsOK() && !status.IsOK())
+    task_status_ = status;
+}
+
+StreamExecutionContext ::~StreamExecutionContext() {}
+
+void StreamExecutionContext ::RecycleNodeInputs(onnxruntime::NodeIndex node_index) {
+  auto* execution_plan = session_state_->GetExecutionPlan();
+  for (auto idx : execution_plan->node_release_list[node_index]) {
+    if (--release_plan_[idx] == 0) {
+      ORT_ENFORCE(frame_.ReleaseMLValue(static_cast<int>(execution_plan->release_actions[idx].value_index)).IsOK());
+      LOGS(*logger_, INFO) << "ort value " << execution_plan->release_actions[idx].value_index << " released";
+    }
+  }
+}
+
+void RunSince(size_t stream_idx, StreamExecutionContext& ctx, SessionScope& session_scope, const bool& terminate_flag, size_t since) {
+  if (!ctx.TaskStatus().IsOK()) {
+    // already in bad status, terminate it
+    ctx.CompleteTask();
+    return;
+  }
+
+  // get logic stream
+  auto& execution_plan = ctx.GetSessionState().GetExecutionPlan()->execution_plan;
+  auto& logic_stream = execution_plan[stream_idx];
+  size_t end = logic_stream->steps_.size();
+#ifdef ENABLE_TRAINING
+  auto* range = ctx.GetCurrentRange();
+  if (range)
+    end = std::min(end, range->stream_pc_range[stream_idx].second);
+#endif
+
+#ifdef ENABLE_TRAINING
+  // this is a special handle for training
+  // with ORTModule, we are partially execute the graph with a shared context.
+  // there is a case that in forward pass we want to trigger downstream which
+  // not in current range. We need to execute one step to consume the Barrier
```

Comment:
>// not in current range [](http://example.com/codeflow?start=2&length=23)

is this the case where some node (Shape for example) after YielOp in topo order is executed in fw pass exectuion?

Reply:
it is more complicated than that... basically if one of the kernel need to notify its consumer, but the consumer is in bw graph, we need to trigger the consumer in fw otherwise the backward pass will hang. In short, the current index-based-range definition is not suitable for multi-stream execution mode.